bmips: update ethernet driver

spliew89 opened the pull request:
Improve performance and add features to the bmips ethernet driver to bring it to
parity with bcm63xx.
- don't use NET_IP_ALIGN in copybreak which messes up alignment with DSA tag
- convert to napi_build_skb()
- use rx batch processing
- add BQL support

Tested on BCM6328 using iperf3.
The use of DSA in bmips actually lowered throughput to ~47Mbps from the max
100Mbps under bcm63xx. This series of changes uplift performance by ~20% to
~54Mbps (with flow offload ~74Mbps).

spliew89 committed:
bmips: use netdev_alloc_skb() for copybreak

bmips is using Broadcom B53 DSA driver which means the ethernet driver
must compensate for 6 bytes tags from the internal switch.
This means using NET_IP_ALIGN actually misaligns the skb, lowering
performance significantly. Therefore napi_alloc_skb() which uses
NET_IP_ALIGN is changed to netdev_alloc_skb().
Performance in iperf3 is increased from ~47Mbps to 52Mbps.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
bmips: convert to build_skb()

We can increase the efficiency of rx path by using buffers to receive
packets then build SKBs around them just before passing into the network
stack. In contrast, preallocating SKBs too early reduces CPU cache
efficiency.
Performance is slightly increased but the changes allow more
potential optimizations.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
bmips: check NAPI context when refilling rx SKB

Check if we're in NAPI context when refilling rx. Normally we're almost
always running in NAPI context. Dispatch to napi_alloc_frag() directly
instead of relying on netdev_alloc_frag() which does the same but with
the overhead of local_bh_disable/enable.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
bmips: switch to napi_build_skb() to reuse skbuff_heads

napi_build_skb() reuses NAPI skbuff_head cache in order to save some
cycles on freeing/allocating skbuff_heads on every new rx or completed
tx.
Use napi_consume_skb() to feed the cache with skbuff_heads of completed
tx so it's never empty.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
bmips: improve rx loop

Use existing rx processed count to track against budget, thereby making
budget decrement operation redundant.
rx_desc_count can be calculated outside the rx loop, making the loop a
bit smaller.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
bmips: batch process rx path

Use netif_receive_skb_list() to batch process rx skb. This improves
cache efficiency.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
bmips: batch process eth_type_trans() in rx path

Improve cache efficiency by running eth_type_trans() in its own loop.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
bmips: add BQL support

Add Byte Queue Limits support to reduce/remove bufferbloat.

Signed-off-by: Sieng Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
bmips: fix Tx cleanup when NAPI poll budget is zero

NAPI poll() function may be passed a budget value of zero, i.e. during
netpoll, which isn't NAPI context.
Therefore, napi_consume_skb() must be given budget value instead of
!force to truly discern netpoll-like scenarios.

https://lore.kernel.org/netdev/20220707141056.2644-1-liew.s.piaw@gmail.com/t/#m470f5c20225e76fb08c44d6cfa2f1b739ffaaea4
Signed-off-by: Sieng-Piaw Liew <liew.s.piaw@gmail.com>

Ansuel commented on target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:
```diff
@@ -251,32 +254,35 @@ static inline void dmas_writel(struct bcm6368_enetsw *priv, u32 val,
+		if (!priv->rx_buf[desc_idx]) {
+			unsigned char *buf;
+
+			if(likely(napi_mode))
```
nitpick space after if.

spliew89 replied:
Done

Ansuel replied:
Mhhh think you misunderstood :D
Was referring to
```
if(likely(napi_mode))
```
that should be changed to
```
if (likely(napi_mode))
```

not adding an extra space after
```
if (!priv->rx_buf[desc_idx]) {
```
(that was fine)

spliew89 replied:
Done

Ansuel commented on target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:
```diff
@@ -251,32 +254,35 @@ static inline void dmas_writel(struct bcm6368_enetsw *priv, u32 val,
+			if(likely(napi_mode))
+				buf = napi_alloc_frag(priv->rx_frag_size);
+			else
+				buf = netdev_alloc_frag(priv->rx_frag_size);
```
nitpick extra space after else condition ending

spliew89 replied:
Done

Ansuel commented on target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:
```diff
@@ -251,32 +254,35 @@ static inline void dmas_writel(struct bcm6368_enetsw *priv, u32 val,
-			desc->address = p;
+			priv->rx_buf[desc_idx] = buf;
+			desc->address = dma_map_single(&priv->pdev->dev,
+										   buf + NET_SKB_PAD,
```
what happen here with the spacing?

spliew89 replied:
Sorry, vscode messes up the spacing sometimes.

The bcm63xx_net driver, which this driver is based on, doesn't check the dma addr.
I can add dma_mapping_error() checking if needed.

Ansuel replied:
yes pls add the dma check
Np about the space it was just to warn you on what to fix :D

Ansuel replied:
(if you want you can put the dma_mapping_error() in a separate commit)

spliew89 replied:
> (if you want you can put the dma_mapping_error() in a separate commit)

https://github.com/openwrt/openwrt/pull/10144/commits/5cf4fccdecc8f980e3bf3284781d2e013e4a0dba

spliew89 replied:
Done

Ansuel commented on target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:
```diff
@@ -365,43 +375,57 @@ static int bcm6368_enetsw_receive_queue(struct net_device *dev, int budget)
-			nskb = napi_alloc_skb(&priv->napi, len);
-			if (!nskb) {
+			unsigned int nfrag_size = SKB_DATA_ALIGN(NET_SKB_PAD + len +
+													 SKB_DATA_ALIGN(sizeof(struct skb_shared_info)));
```
extra spacing?

spliew89 replied:
Done

Ansuel commented on target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:
```diff
@@ -365,43 +375,57 @@ static int bcm6368_enetsw_receive_queue(struct net_device *dev, int budget)
-					 priv->rx_skb_size, DMA_FROM_DEVICE);
-			priv->rx_skb[desc_idx] = NULL;
+			dma_unmap_single(kdev, desc->address,
+							 priv->rx_buf_size, DMA_FROM_DEVICE);
```
wrong spacing?

spliew89 replied:
Done

Ansuel commented on target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:
```diff
@@ -960,9 +992,12 @@ static int bcm6368_enetsw_probe(struct platform_device *pdev)
 				  priv->dma_maxburst * 4);
 
+	priv->rx_frag_size = SKB_DATA_ALIGN(NET_SKB_PAD + priv->rx_buf_size +
+										SKB_DATA_ALIGN(sizeof(struct skb_shared_info)));
```
wrong spacing?

spliew89 replied:
Done

Ansuel commented:
@spliew89 thanks for the changes but you should fix the related commits instead of creating an additional commit that fix the spacing.
(so use git rebase, edit the related commit with the wrong spacing and then force push)

spliew89 committed:
bcm6368-enetsw: add and fix spacing

Add and fix spacing as requested by Ansuel.

Signed-off-by: Sieng-Piaw Liew <liew.s.piaw@gmail.com>

spliew89 committed:
bmips: check for DMA error when refilling Rx

dma_mapping_error() should be called as dma_map_single() could fail and
return error.

Signed-off-by: Sieng-Piaw Liew <liew.s.piaw@gmail.com>

Noltari commented:
Merged, thanks @spliew89!
https://github.com/Noltari/openwrt/compare/3817577c92618256565a55213e440cbcd1bd4e86...65ebe311304a32423a46bf4efe62f29cda06acab
